### PR TITLE
Write index metadata with a new version 

### PR DESCRIPTION
**Pull request by eddyxu**
* IndexBuilder / WritePath write index section in the manifest file.
* Use uuid to identify an index cross dataset versions
* `Scanner::into_stream` will choose the KNN(Index) node if the column exists in the index section in manifest.

### CODE REVIEW

**Comment by eddyxu** [comment]
C++ / Notebook failure because C++ arrow upgrading upstream. will try to fix them later in a few weeks.

File: protos/format.proto [hunk 1]
```diff
@@ -58,6 +63,9 @@ message Manifest {
 
   // Schema metadata.
   map<string, bytes> metadata = 5;
+
+  // If presented, the file position of the index metadata.
+  optional uint64 index_section = 6;
```
Comment (changhiskhan) [inline protos/format.proto:68]: So proto2 only?

File: rust/src/dataset.rs [hunk 2]
```diff
@@ -216,6 +213,85 @@ impl Dataset {
...
+    /// Parameters:
+    ///
+    ///  - `columns`: the columns to build the indices on.
+    ///  - `index_type`: specify [`IndexType`].
+    ///  - `name`: optional index name. Must be unique in the dataset.
+    ///            if not provided, it will auto-generate one.
+    ///  - `params`: index parameters.
+    pub async fn create_index(
+        &self,
+        columns: &[&str],
```
Comment (changhiskhan) [inline rust/src/dataset.rs:229]: Nice this leaves room for combined search

File: rust/src/dataset.rs [hunk 3]
```diff
@@ -319,14 +407,53 @@ impl Dataset {
     pub fn fragments(&self) -> &Arc<Vec<Fragment>> {
         &self.manifest.fragments
     }
+
+    /// Read all indices of this Dataset version.
+    pub async fn load_indices(&self) -> Result<Vec<Index>> {
```
Comment (changhiskhan) [inline rust/src/dataset.rs:412]: So each scan has to re-read the indices? Would it make sense to just keep it in memory ?

File: rust/src/index/vector/kmeans.rs [hunk 4]
```diff
@@ -32,12 +32,17 @@ fn train_kmeans_faiss(
...
-/// Fallback implementation of KMeans.
-fn train_kmeans_fallback() -> Result<Float32Array> {
+/// A fallback default implementation of KMeans, if not accelerator found.
+fn train_kmeans_fallback(
+    _array: &Float32Array,
+    _dimension: usize,
+    _k: u32,
+    _max_iter: u32,
+) -> Result<Float32Array> {
     todo!()
```
Comment (changhiskhan) [inline rust/src/index/vector/kmeans.rs:42]: Should this panic just so it doesn't pass silently?

**Review by changhiskhan — COMMENTED** [review]
Few minor questions.

File: protos/format.proto (hunk 1 above)
Comment (eddyxu) [reply protos/format.proto:68]: TIL that `proto3` treats optional as explicitly presence https://protobuf.dev/programming-guides/field_presence/

File: rust/src/dataset.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/dataset.rs:412]: That makes sense. Let's keep that as future work? So is caching the IVF partitions, assuming there will be strong locality during queries.

File: rust/src/index/vector/kmeans.rs (hunk 4 above)
Comment (eddyxu) [reply rust/src/index/vector/kmeans.rs:42]: oh, `todo()` actually raises `not implemented` error. that's why I disabled the test in `Dataset.rs`, unfortunately.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**